src/pyspiffe/spiffe_id/spiffe_id.py

```diff
@@ -62,114 +151,130 @@ def of(
-        return '{}://{}'.format(SPIFFE_SCHEME, self.__trust_domain.name())
-
-    # path can be an array of path segments or a single string representing a path
-    def __set_path(self, path: Union[str, List[str]]) -> None:
-        if isinstance(path, list):
-            self.__path = ''
-            for s in path:
-                self.__path += self.normalize_path(s)
-            return
-        self.__path = self.normalize_path(path)
-
-    def __set_trust_domain(self, trust_domain: TrustDomain) -> None:
-        self.__trust_domain = trust_domain
+        return '{}://{}'.format(SPIFFE_SCHEME, self._trust_domain.name())
 
     def path(self) -> str:
-        return self.__path
+        return self._path
 
     def trust_domain(self) -> TrustDomain:
-        return self.__trust_domain
+        return self._trust_domain
 
     def is_member_of(self, trust_domain: TrustDomain) -> bool:
-        return self.__trust_domain == trust_domain
-
-    @staticmethod
-    def parse_and_validate_uri(spiffe_id: str) -> Dict:
-        if len(spiffe_id) > SPIFFE_ID_MAXIMUM_LENGTH:
-            raise ArgumentError(
-                'SPIFFE ID: maximum length is {} bytes'.format(SPIFFE_ID_MAXIMUM_LENGTH)
-            )
-        try:
-            uri = parse(spiffe_id.strip(), rule='URI')
-        except ValueError as err:
-            raise ArgumentError(str(err))
-
-        scheme = uri.get('scheme')
-        if scheme.lower() != SPIFFE_SCHEME:
-            raise ArgumentError('SPIFFE ID: invalid scheme: expected spiffe')
-
-        query = uri.get('query')
-        if query:
-            raise ArgumentError('SPIFFE ID: query is not allowed')
-
-        fragment = uri.get('fragment')
-        if fragment:
-            raise ArgumentError('SPIFFE ID: fragment is not allowed')
-
-        authority = uri.get('authority')
-        if not authority:
-            raise ArgumentError('SPIFFE ID: trust domain cannot be empty')
-
-        # has user@info:
-        if '@' in authority:
-            raise ArgumentError('SPIFFE ID: userinfo is not allowed')
-
-        # has domain:port
-        if ':' in authority:
-            raise ArgumentError('SPIFFE ID: port is not allowed')
-
-        return uri
-
-    @staticmethod
-    def normalize_path(path: str) -> str:
-        path = path.strip()
-        if path and not path.startswith('/'):
-            return '/' + path
-        return path
+        return self._trust_domain == trust_domain
+
+    def _set_trust_domain(self, trust_domain: TrustDomain) -> None:
+        self._trust_domain = trust_domain
+
+    def _set_path(self, path: str):
+        self._path = path
+
+
+def validate_trust_domain_name(name):
+    for c in name:
+        if not is_valid_trustdomain_char(c):
+            raise SpiffeIdError(BAD_TRUST_DOMAIN_CHAR)
+
+
+def is_valid_trustdomain_char(c) -> bool:
+    if 'a' <= c <= 'z':
+        return True
+    if '0' <= c <= '9':
+        return True
+    return c == '-' or c == '.' or c == '_'
```

Comment:
Is there a way to define those ranges of characters without needing to put each character in the list ?

Reply:
Not that I can tell you now. I was about to look for it but then I saw on the test file that the list of characters is already created. I think is can be copied from there.

Reply:
well in that case I'd rather prefer to keep the `if` condition as it is, that checks that the given char is in one of the ranges because otherwise listing all characters is prone to errors and I prefer to just use it in the tests.